app/src/main/java/com/asfoundation/wallet/topup/amazonPay/AmazonPayTopUpViewModel.kt

```diff
@@ -0,0 +1,197 @@
+  private val createAmazonPayTransactionTopUpUseCase: CreateAmazonPayTransactionTopUpUseCase,
+  private val getAmazonPayCheckoutSessionIdUseCase: GetAmazonPayCheckoutSessionIdUseCase,
+  private val getAmazonPayChargePermissionUseCase: GetAmazonPayChargePermissionUseCase,
+  private val saveAmazonPayChargePermissionLocalStorageUseCase: SaveAmazonPayChargePermissionLocalStorageUseCase,
+  private val getAmazonPayChargePermissionLocalStorageUseCase: GetAmazonPayChargePermissionLocalStorageUseCase,
+  private val patchAmazonPayCheckoutSessionUseCase: PatchAmazonPayCheckoutSessionUseCase,
+  private val displayChatUseCase: DisplayChatUseCase,
+  private val analytics: TopUpAnalytics
+) : ViewModel() {
+
+  lateinit var paymentData: TopUpPaymentData
+  private val _uiState = MutableStateFlow<UiState>(UiState.Loading)
+  var uiState: StateFlow<UiState> = _uiState
+  var amazonTransaction: AmazonPayTransaction? = null
+  private val JOB_UPDATE_INTERVAL_MS = 5 * DateUtils.SECOND_IN_MILLIS
+  private val JOB_TIMEOUT_MS = 600 * DateUtils.SECOND_IN_MILLIS
+  private var jobTransactionStatus: Job? = null
+  private val timerTransactionStatus = Timer()
+  private var isTimerRunning = false
+  val scope = CoroutineScope(Dispatchers.Main)
+  var runningCustomTab = false
+
+  @SuppressLint("CheckResult")
+  fun getPaymentLink() {
+    getAmazonPayChargePermissionUseCase()
+      .flatMap { chargePermissionId ->
+        saveAmazonPayChargePermissionLocalStorageUseCase(chargePermissionId = chargePermissionId.chargePermissionId)
+        createAmazonPayTransaction(chargePermissionId.chargePermissionId)
+      }
+      .onErrorResumeNext {
+        createAmazonPayTransaction(null)
+      }
+      .doOnSubscribe { _uiState.value = UiState.Loading }
+      .doOnSuccess { amazonTransactionResult ->
+        validateResultOfPaymentLink(amazonTransactionResult)
+      }
+      .subscribe({}, { _ -> _uiState.value = UiState.Error })
+  }
+
+  private fun createAmazonPayTransaction(chargePermissionId: String?): Single<AmazonPayTransaction> {
+
+    val price = AmazonPrice(value = paymentData.fiatValue, currency = paymentData.fiatCurrencyCode)
+    return createAmazonPayTransactionTopUpUseCase(
+      price = price,
+      chargePermissionId = chargePermissionId
+    )
+  }
+
+  private fun validateResultOfPaymentLink(amazonPayTransaction: AmazonPayTransaction) {
+    amazonTransaction = amazonPayTransaction
+    when {
+      amazonPayTransaction.errorCode == null && !amazonPayTransaction.redirectUrl.isNullOrEmpty()  ->
+        _uiState.value = UiState.PaymentRedirect3ds
+      amazonPayTransaction.errorCode == null && getAmazonPayChargePermissionLocalStorageUseCase().isEmpty() ->
+        _uiState.value = UiState.PaymentLinkSuccess
+      amazonPayTransaction.errorCode == null && getAmazonPayChargePermissionLocalStorageUseCase().isNotEmpty() ->
+        startTransactionStatusTimer()
+
+      else ->
+        _uiState.value = UiState.Error
```

Comment:
Just a question, if the chargePermissionId fails the payment (for example, if the user has insufficient balance) would the user get stuck with a always failing payment? If this is the case, it's better to redirect the user to login from zero again.